src/ThreadPool/MyThreadPool.cs

```diff
@@ -0,0 +1,341 @@
+namespace ThreadPool;
+
+using Extensions;
+using System.Collections.Concurrent;
+using Optional;
+using Exceptions;
+
+/// <summary>
+/// Thread pool that implement <see cref="IDisposable"/>.
+/// </summary>
+public sealed class MyThreadPool : IDisposable
+{
+    private readonly BlockingCollection<Action> queue;
+
+    private readonly CountdownEvent threadsCompletedEvent;
+
+    private readonly object locker = new ();
+
+    private volatile bool isDisposed;
+    private volatile bool isShutDown;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="MyThreadPool"/> class.
+    /// </summary>
+    /// <param name="threadCount">
+    /// Number strictly greater than zero - number of threads processing tasks.
+    /// </param>
+    public MyThreadPool(int threadCount)
+    {
+        if (threadCount <= 0)
+        {
+            throw new MyThreadPoolException(
+                $"the number of threads must be greater than 0, but was = {threadCount}");
+        }
+
+        this.queue = new BlockingCollection<Action>();
+
+        this.threadsCompletedEvent = new CountdownEvent(threadCount);
+
+        for (var i = 0; i < threadCount; i++)
+        {
+            new ThreadPoolItem(this.queue, this.threadsCompletedEvent).Ignore();
+        }
+    }
+
+    /// <summary>
+    /// A method that allows you to add a task for execution in the form of <see cref="Func{TResult}"/>.
+    /// </summary>
+    /// <param name="func">Function whose value will be computed as one.</param>
+    /// <typeparam name="TResult">Function result type.</typeparam>
+    /// <returns>Abstraction over the task accepted for execution, see <see cref="IMyTask{TResult}"/>.</returns>
+    public MyTask<TResult> Submit<TResult>(Func<TResult> func)
+    {
+        var resultTask = Option.None<MyTask<TResult>>();
+
+        try
+        {
+            Monitor.Enter(this.locker);
+
+            if (!this.isShutDown)
+            {
+                var (newTask, newCell) = CreateNewTaskAndCell(func, this);
+
+                resultTask = newTask.Some<MyTask<TResult>>();

```

Comment:
```suggestion
                resultTask = newTask.Some();

```
Параметры-типы методов компилятор сам выводит

Reply:
Проект без явной спецификации типа не собирается, возможно я что-то не понимаю...

```
error CS1929: 'MyThreadPool.MyTask<TResult>' does not contain a definition for 'Some' and the best e
xtension method overload 'OptionExtensions.Some<?>(?)' requires a receiver of type '?'
error CS8389: Omitting the type argument is not allowed in the current context
```